src/libponyc/ast/ast.c

```diff
@@ -457,9 +465,68 @@ ast_t* ast_dup(ast_t* ast)
   return duplicate(NULL, ast);
 }
 
+ast_t* ast_dup_partial(ast_t* ast, bool* dup_child, bool dup_type,
+  bool dup_annotation, bool dup_symtab)
```

Comment:
Should `ast_dup_partial` enforce frozen-ness with an assert?

Reply:
Since this function (and the existing `ast_dup`) creates a deep copy of the AST, it shouldn't be needed.

Reply:
I was a bit confused at what this was doing - I thought the non-duped elements were being kept on in the structure without changes, but I see now that you're replacing them with `ast_blank`.